mesonbuild/options.py

```diff
@@ -682,96 +704,370 @@ def add_to_argparse(self, name: str, parser: argparse.ArgumentParser, help_suffi
+
+        This way everyone can do f.ex, get_option('libdir') and usually get
+        the library directory relative to prefix, even though it really
+        should not be relied upon.
+        '''
+        try:
+            value = pathlib.PurePath(value)
+        except TypeError:
+            return value
+        if option.name.endswith('dir') and value.is_absolute() and \
+           option not in BUILTIN_DIR_NOPREFIX_OPTIONS:
+            try:
+                # Try to relativize the path.
+                value = value.relative_to(prefix)
+            except ValueError:
+                # Path is not relative, let’s keep it as is.
+                pass
+            if '..' in value.parts:
+                raise MesonException(
+                    f'The value of the \'{option}\' option is \'{value}\' but '
+                    'directory options are not allowed to contain \'..\'.\n'
+                    f'If you need a path outside of the {prefix!r} prefix, '
+                    'please use an absolute path.'
+                )
+        # .as_posix() keeps the posix-like file separators Meson uses.
+        return value.as_posix()
+
+    def set_value(self, key: T.Union[OptionKey, str], new_value: 'T.Any', first_invocation: bool = False) -> bool:
+        key = self.ensure_and_validate_key(key)
+        if key.name == 'prefix':
+            new_value = self.sanitize_prefix(new_value)
+        elif self.is_builtin_option(key):
+            prefix = self.get_value_for('prefix')
+            new_value = self.sanitize_dir_option_value(prefix, key, new_value)
+        if key not in self.options:
+            raise MesonException(f'Unknown options: "{key.name}" not found.')
+
+        valobj = self.options[key]
+        old_value = valobj.value
+        changed = valobj.set_value(new_value)
+
+        if valobj.readonly and changed and not first_invocation:
+            raise MesonException(f'Tried to modify read only option {str(key)!r}')
+
+        if key.name == 'prefix' and first_invocation and changed:
+            self.reset_prefixed_options(old_value, new_value)
+
+        if changed:
+            self.set_dependents(key, new_value)
+
+        return changed
 
     # FIXME, this should be removed.or renamed to "change_type_of_existing_object" or something like that
     def set_value_object(self, key: T.Union[OptionKey, str], new_object: 'UserOption[T.Any]') -> None:
-        key = self.ensure_key(key)
-        self.d[key] = new_object
+        key = self.ensure_and_validate_key(key)
+        self.options[key] = new_object
+
+    def set_dependents(self, key: OptionKey, value: str):
+        if key.name != 'buildtype':
+            return
+        opt, debug = {'plain': ('plain', False),
+                      'debug': ('0', True),
+                      'debugoptimized': ('2', True),
+                      'release': ('3', False),
+                      'minsize': ('s', True),
+                      }[value]
```

Comment:
Putting this dictionary in the class as a classvar constant will both make this more readable, and allow the bytecode compiler to do some better optimization.